src/sandbox/algorithms/dqn/dqn.py

```diff
@@ -1,42 +1,73 @@
+from __future__ import annotations
+import logging
+from sandbox.action_selection_rules.generic import ActionSelectionRule
+from sandbox.policies.nn_policy import QNetwork
+from sandbox.algorithms.algorithm import Algorithm
 from sandbox.algorithms.dqn.replay_buffer import ReplayMemory, Transition
-from sandbox.algorithms.dqn.policy import QNetwork
+
 
 import gym
 import gym.spaces
-from abc import ABC
-from typing import Callable
+from typing import Callable, TypeVar
 import numpy as np
 import tensorflow as tf
 
+from sandbox.wrappers.discrete_env_wrapper import DiscreteEnvironment
+
 
-class DQNAlgorithm:
+QNetworkType = TypeVar('QNetworkType', bound=QNetwork)
+class DQNAlgorithm(Algorithm[np.ndarray, int, QNetworkType]):
     def __init__(
         self,
-        env: gym.Env,
         memory_size: int,
-        network: type[QNetwork],
-        policy: Callable,
+        create_network: Callable[[gym.spaces.Box, gym.spaces.Discrete, float], QNetworkType],
```

Comment:
Zmieniłem sygnaturę, bo może być tak, że ktoś napisze sieć która wymaga więcej argumentów niż te trzy które my przekazujemy i teraz dzięki temu że jest wymagana funkcja, a nie podtyp będzie w stanie to zrobić.